Major CLI documentation overhaul and automation

lukpueh opened the pull request:
**Fixes issue #**:
Partially fixes #277 and #239 
Supersedes #332 

**Description of the changes being introduced by the pull request**:

This PR performs a major revision of all in-toto cli tools so that the description-, help and usage texts can be used as single source of truth, both in the command line (e.g. when using the `--help` flag on any of the `in-toto-*` commands, and when auto-generating sphinx documentation (e.g. for readthedocs). 

Changes include:
- refactor cli modules and enhance contents and format of help texts (8580862, cd30f84, eccedc8, 881ca1e, ef48a7e, 6ef8f95)
- add custom sphinx argparse extension and docs generation configuration (ed2dec8, ea5cf92)
- streamline cli module document headers and update docs links in main in-toto README.md (3b5c04f, 5578d3e)

Apologies for the size of the diff! I Strongly suggest to review commit by commit, reading the commit messages. I also suggest to go through the rendered documentation on readtheodcs:
https://in-toto-lukpueh.readthedocs.io/en/latest/command-line-tools/index.html

Big kudos to @SantiagoTorres and his work in #332, which I have integrated in this PR.


**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Add custom sphinx argparse-example-epilog extension

The in_toto command line tools use argparse epilog to add an
example usage section to the help message. The epilog does not
render nicely when using the sphinx-argparse extension.

This commit adds a custom extension to properly render the title
(expected to be 'EXAMPLE USAGE') and code blocks of a command line
tools epilog.

See ArgparseUsageEpilog docstring for details about the extension.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Add argparse action group helper functions

Default argparse action groups are "optional arguments" and
"positional arguments". Most in-toto command line tools also use a
custom "required named arguments" action group.

This commit adds two helper function, one to make all group titles
title case, e.g. for consistency when using the sphinx argparse
extension, which title-cases default action groups only; and one
function to customize the order in which action groups are
displayed in the help message or in sphinx, to e.g. show required
arguments before optional arguments.

Note that the helpers access protected members of argparse.
Although not part of the official API these seem pretty stable, and
we don't heavily rely on them, in case they disappear, unit tests
will catch it and we can remove/replace the feature, for which they
are currently used.

This commit also adds tests for the two helpers.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Move argparser creation to separate function

Separate argument parser creation and configuration into a
create_parser function in all in-toto command line tools. This is
required to auto-generate sphinx documentation.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Customize cli action group titles and order

Call helper functions to make argparse action group titles tile
case and to create a action group order for all in-toto cli tools.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Enhance CLI descriptions, help and usage examples

Major overhaul of all CLI help texts, in an attempt to make them
more informative and render nicely and uniformly on
sphinx/readthedocs.

Kudos to @SantiagoTorres and his work in
https://github.com/in-toto/in-toto/pull/332, which I've been
integrating here.

Changes includes:

- add more usage examples to arpgarse.epilog field of cli tools
- Unify title and indentation in all usage examples to render nicely
  with the custom argparse-example-epilog sphinx extension
- uniformly start option help texts with lowercase
- uniformly start options with arguments with a noun (the argument)
  and options without arguments with a verb (what the option does)
- unify argument groups, i.e. use defaults (positional arguments,
  optional arguments) and one custom (required named arguments)
  where applicable
- uniformly use single quotes instead of backticks for emphasis in
  help texts
- move more common arguments (must be exactly the same!) to
  common_args.py

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Add sphinx auto cli doc configuration

Use sphinx-argparse and custom argparse-epilog sphinx extensions
to auto-generate cli documentation.

The directives provided by the two extensions are used in the newly
created doc/source/command-line-tools/in-toto-*.rst files, and
insert the output of the corresponding cli tool's help messages when
generating documentation with sphinx.

This commit further
- updates the table of context,
- adds a docs specific requirements file to install the 3rd-party
  extension)
- patches the path to make available the custom extension

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Add requirements file to build docs

Sphinx installs in-toto with its runtime dependencies, and
extra packages, e.g. the sphinx-argparse.

This file should be chosen via the readthedocs dashboard.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Manually check mutual exclusion in in-toto-record

Switch to manual mutual exclusion check for --key and --gpg
arguments (as in in_toto_run).

The `add_mutually_exclusive_group` used for the --key and --gpg
options can not be called on a custom argument group, such as the
"required named arguments" group. As a consequence the options
are listed in the "optional arguments" section. Since at least one
of them is required it is better to list both of them in the
"required named arguments" group (in addition to clarifying their
use in via the per-argument description.

This commit removes the use of add_mutually_exclusive_group and
siwtches to manual mutual exclusion check for --key and --gpg
arguments (as in in_toto_run) for better help message/cli docs.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Disambiguate gpg home dir and keyring

Some functions in securesystemslib.gpg support a homedir argument.
This corresponds to the gpg `--homedir` option, which is different
from the gpg `--keyring` option. However, in-toto documentation
inaccurately refers to the `homedir` as keyring in some places.
This commit fixes the relevant parts in the code and cli docs.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Remove help text from cli doc headers

Remove output of calling cli tools with the --help flag from all
cli modules' document headers as it needs to be copied/updated
manually and thus is prone to get out of sync.

The same (single-sourced) information can be found
- in the code,
- by installing the in-toto and running the cli command with
  the --help flag, and soon
- on in-toto.readthedocs.io

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Update cli documentation links in README.md

Point to readthedocs pages instead of document header of the
individual CLI modules.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

SantiagoTorres commented on tests/test_common_args.py:
```diff
@@ -0,0 +1,78 @@
+    # NOTE: argparse could remove this at any time without notice
+    self.assertTrue(
+        isinstance(getattr(self.parser, "_action_groups", None), list))
+
```
I like we are testing against this, but I've always been wary of `type` vs `isinstance`. I personally don't know if we want to ensure it's a list-ish type, or must be a list and nothing else. What's your take?

lukpueh replied:
You are absolutely right. In this case we should do a `type` and not an `instance` check, because in `common_args.sort_action_groups` we replace what `_action_groups` references with a list.  So if `argparse` were to change the type to something list-ish, it could lead to unexpected behavior and we might not detect it in the test here. Thanks for pointing this out. I'll push a change.

SantiagoTorres commented on in_toto/util.py:
```diff
@@ -186,8 +186,8 @@ def import_public_keys_from_files_as_dict(filepaths, key_types=None):
-  identified by the list of passed `keyids` from the gpg keyring at `gpg_home`.
-  If `gpg_home` is False the default keyring is used. """
+  identified by the list of passed `keyids` from a gpg keyring in `gpg_home`.
+  If `gpg_home` is False the default gpg home directory is used. """
```
Thanks for disambiguating this. It makes me wonder whether we want to use `--keyring` instead later down the line though. What do you think?

lukpueh replied:
I wouldn't replace it because there is merit to overriding the homedir and we do this in tests a lot. But it would definitely be nice to add support for the`--keyring` option. Maybe worth a feature request in securesystemslib?

SantiagoTorres commented on README.md:
```diff
@@ -133,7 +136,8 @@ formats or signing routines change.
 `in-toto-sign --help` or look at the
-[code documentation](https://github.com/in-toto/in-toto/blob/develop/in_toto/in_toto_sign.py).
+[online
+documentation](https://in-toto.readthedocs.io/en/latest/command-line-tools/in-toto-sign.html).
```
Is it me, or the line wrapping here also causes the links to be split? (or it could be the richdiff viewer being wonky as well..)

lukpueh replied:
AFAICS the markdown parser can deal line breaks inside brackets. The rendered README also suggests that that's the case: https://github.com/in-toto/in-toto/blob/5578d3e871171cc1f9bcce9f4228a58c2ab7ef27/README.md

SantiagoTorres commented:
Mostly minor nits, although I'd like to bring some of these to discussion. Great work!!

Commit:
Check type and not instance in test_common_args

When testing the undocumented (protected) attribute
'_action_groups' from 'argparse', it is safer to do a type and not
an instance check, because in 'common_args.sort_action_groups' we
replace what '_action_groups' references with a list. So if argparse
were to change its type to something list-ish, it could lead to
unexpected behavior on our side and we might not detect it in the
test here.

Kudos to @SantiagoTorres for pointing this out.

Pylint claims that this is an unidiomatic-typecheck. We beg to
differ.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Change order of argument sections in CLI docs

Change the section order of argument types to show "position
arguments" (required) before "optional arguments" (not required).

Further update in_toto_record to make the ordering and naming
consistent, following advice from @jhdalek55.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Refine install and CLI docs

Reinfe CLI and installation docs as shown on readthedocs following
an out-of-band review by @jhdalek55.

This mostly adds missing preposition and fixes wording.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

SantiagoTorres commented on doc/source/installing.rst:
```diff
@@ -38,22 +38,22 @@ To install using pip, simply run:
-- `GPG <https://gnupg.org>`_ if you plan on generating gpg signatures (verifying
-  works without gpg).
+- `OpenSSL <https://openssl.org>`_ used to generate and verify RSA signatures, and to export and verify signatures created with GPG.
+- `GPG <https://gnupg.org>`_ if you plan on generating GPG signatures (verifying
```
I wonder if we'd like to say `if you plan on generating PGP signatures` instead.

lukpueh replied:
Or we say `[PG]{3}` and let the reader decide, additionally we should add a reference to [RFC 4880 -- 1.1. Terms](https://tools.ietf.org/html/rfc4880#section-1.1) :)

SantiagoTorres commented on doc/source/installing.rst:
```diff
@@ -38,22 +38,22 @@ To install using pip, simply run:
 ----------------------
 
-If your system doesn't provide in-toto yet, you can also install it from the
+If your system doesn't yet provide in-toto, you can also install it from the
```
I'm not completely convinced moving this yet is necessary? I'm also wondering if we could simplify this:

```
if your system doesn't provide in-toto, you can install it from the source
```

to be a little bit more succint.

SantiagoTorres reviewed (commented):
I found two minor nits (that I'm not completely sure are worth fixing). I can merge afterwards. Thanks for this!

Commit:
Polish installation instructions

Following @SantiagoTorres' review, this commit updates the
installation instructions to:
- disambiguate GPG (the tool) and PGP (the standard), and
- shorten another phrase.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

lukpueh commented:
Thanks for the feedback, @SantiagoTorres. I just pushed a commit that incorporates both of your suggestions.

SantiagoTorres commented:
Thank you!
